discord/guild_builder.py

```diff
@@ -0,0 +1,788 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Coroutine, Generator, overload
+
+from . import utils
+from .abc import Snowflake
+from .colour import Color, Colour
+from .emoji import Emoji, PartialEmoji
+from .enums import ChannelType, VoiceRegion
+from .flags import SystemChannelFlags
+from .guild import Guild
+from .permissions import PermissionOverwrite, Permissions
+from .state import ConnectionState
+
+if TYPE_CHECKING:
+    from .types.guild import GuildCreate as GuildCreatePayload
+    from .types.role import Role as RolePayload
+
+MISSING = utils.MISSING
+
+__all__ = (
+    "GuildBuilder",
+    "GuildBuilderChannel",
+    "GuildBuilderRole",
+)
+
+
+class GuildBuilder:
+    """Represents a Discord guild that is yet to be created.
+
+    This is returned by :meth:`Client.create_guild` to allow users to modify the guild
+    properties before creating it.
+
+    .. versionadded:: 2.7
+
+    .. container:: operations
+
+        .. describe:: await x
+
+            Creates the guild and returns a :class:`Guild` object.
+
+        .. describe:: x()
+
+            Returns a :ref:`coroutine <coroutine>` that, when awaited,
+            returns the new :class:`Guild` object.
+
+    Attributes
+    ----------
+    code: Optional[:class:`str`]
+        The guild template code that the guild is going to be created with.
+    afk_channel_id: Optional[:class:`int`]
+        The AFK channel ID. Defaults to ``None``.
+    system_channel_id: Optional[:class:`int`]
+        The system channel ID. Defaults to ``None``.
+    system_channel_flags: Optional[:class:`SystemChannelFlags`]
+        The system channel flags. Defaults to ``None``. This requires :attr:`.system_channel_id` to be
+        not ``None``.
+    """
+
+    __slots__ = (
+        "_state",
+        "_name",
+        "_icon",
+        "_channels",
+        "_roles",
+        "_metadata",
+        "afk_channel_id",
+        "system_channel_id",
+        "system_channel_flags",
+        "code",
+    )
+
+    def __init__(
+        self,
+        *,
+        state: ConnectionState,
+        name: str,
+        icon: bytes,
+        code: str | None,
+        metadata: dict[str, Any],
+    ) -> None:
+        self._state: ConnectionState = state
+        self._name: str = name
+        self._icon: bytes = icon
+        self._channels: dict[int, GuildBuilderChannel] = {}
+        self._roles: dict[int, GuildBuilderRole] = {}
+        self._metadata: dict[str, Any] = metadata
+        self.code: str | None = code
+        self.afk_channel_id: int | None = None
+        self.system_channel_id: int | None = None
+        self.system_channel_flags: SystemChannelFlags | None = None
+
+    async def _do_create(self) -> Guild:
```

Comment:
This could be public, to allow devs to use it as an alternative to the `__await__` approach, but maybe under another name than `do_create`.

Reply:
Agreed on this, just calling `await GuildBuilder()` seems too ambiguous